nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractCheckHandler.java

```diff
@@ -0,0 +1,131 @@
+import com.uber.nullaway.NullabilityUtil;
+import com.uber.nullaway.Nullness;
+import com.uber.nullaway.handlers.BaseNoOpHandler;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * This Handler parses the jetbrains @Contract annotation and tries to check if the contract is
+ * followed.
+ *
+ * <p>Currently, it supports the case when there is only one clause in the contract. The clause of
+ * the form in which all the elements of the antecedent are either of "_", "null" or "!null", and
+ * the consequent is "!null" is supported. The handler checks and warns under the conditions of the
+ * antecedent if the consequent is "!null" and there is a return statement with "nullable" or "null"
+ * expression.
+ */
+public class ContractCheckHandler extends BaseNoOpHandler {
+
+  private Set<TreePath> supportedMethods = new HashSet<>();
+
+  @Override
+  public void onMatchMethod(
+      NullAway analysis, MethodTree tree, VisitorState state, Symbol.MethodSymbol methodSymbol) {
+    Symbol.MethodSymbol callee = ASTHelpers.getSymbol(tree);
+    Preconditions.checkNotNull(callee);
+    // Check to see if this method has an @Contract annotation
+    String contractString = getContractFromAnnotation(callee);
+
+    if (contractString != null) {
+      // Found a contract, lets parse it.
+      String[] clauses = contractString.split(";");
+      if (clauses.length != 1) {
+        return;
+      }
+
+      String clause = clauses[0];
+      String[] antecedent =
+          getAntecedent(clause, tree, analysis, state, callee, tree.getParameters().size());
+      String consequent = getConsequent(clause, tree, analysis, state, callee);
+
+      boolean supported = true;
+
+      for (int i = 0; i < antecedent.length; ++i) {
+        String valueConstraint = antecedent[i].trim();
+        if (!(valueConstraint.equals("_")
+            || valueConstraint.equals("!null")
+            || valueConstraint.equals("null"))) {
+          supported = false;
+        }
+      }
+
+      if (!consequent.equals("!null")) {
+        supported = false;
+      }
+
+      if (!supported) {
+        return;
+      }
+
+      supportedMethods.add(state.getPath());
+    }
+  }
+
+  @Override
+  public void onMatchReturn(NullAway analysis, ReturnTree tree, VisitorState state) {
+    final TreePath enclosingMethod =
+        NullabilityUtil.findEnclosingMethodOrLambdaOrInitializer(state.getPath());
+    if (!supportedMethods.contains(enclosingMethod)) {
+      return;
+    }
+
+    final Nullness nullness =
+        analysis
+            .getNullnessAnalysis(state)
+            .getNullnessForContractDataflow(
+                new TreePath(state.getPath(), tree.getExpression()), state.context);
+
+    if (nullness == Nullness.NULLABLE || nullness == Nullness.NULL) {
+      reportMatchForContractIssue(
+          tree,
+          "@Contract might not be followed, as returning @Nullable from method with @NonNull return expected in contract",
```

Comment:
Not sure this is a great message. Worth: a) showing the contract that is being violated, and b) if possible, explain the violation.

Something like:

```
"Method @Nullable foo(@Nullable a) has @Contract(!null -> !null), but this appears to be violated, as a @Nullable value may be returned even when a is non-null."
```

Reply:
Let me know how it looks now

Reply:
Don't see a change, actually.

Reply:
I am specifying the contract within the error now and rephrased it a little as you suggested. The snippet above is outdated